Convert to ghc-9.2.1

shayne-fletcher opened the pull request:
This is a complete port of HLint to the ghc-9.2.1 parse tree. All tests pass.

shayne-fletcher commented:
Cc @alanz All done mate.

ndmitchell commented:
Awesome work!

ndmitchell commented:
GHC 9.2.1 is now out. Thoughts on this?

shayne-fletcher commented:
> GHC 9.2.1 is now out. Thoughts on this?

Got it. Time to move on it then. Leave it with me for now and I'll keep you posted.

alanz commented:
For the apply-refact part it will need me to complete my current ghc-exactprint update.  Which is getting close.

ndmitchell commented:
I'm in no tearing rush - a leisurely upgrade is fine :)

shayne-fletcher commented:
i rebased this branch on master today (oct 30, 2021).

the tests all pass but this one in 'Extensions.hs':
```
{-# LANGUAGE DeriveGeneric, TypeFamilies #-} \
data family Bar a; data instance Bar Foo = Foo deriving Generic
```
which produces the apparently unexpected output
```
Note: Extension DeriveGeneric is not used
```
(using the 9.2.1 release candidate ghc-lib-parser-9.2.1.20210509.tar.gz that the `stack.yaml` on this branch arranges to use)

this test was introduced only recently in https://github.com/ndmitchell/hlint/pull/1277. @zliu41 as the author of that test might i trouble you to take a look? i removed the test for the moment. 

this puts us in a good place for the next step: upgrade this branch to the ghc-lib-parser and ghc-lib-parser-ex  newly released 9.2.1 official versions.

shayne-fletcher commented:
fixes issue https://github.com/ndmitchell/hlint/issues/1314

zliu41 commented:
@shayne-fletcher Removing that test for now is good. Once you make cabal build work for this patch, I can look into it - otherwise it's difficult for me to investigate.

LeventErkok commented:
@shayne-fletcher Maybe it's good to go now for GHC 9.2.1?

shayne-fletcher commented:
> @shayne-fletcher Maybe it's good to go now for GHC 9.2.1?

yep, it's high priority with us too. don't worry @LeventErkok , we got you! little patience please 😉

shayne-fletcher commented:
@ndmitchell i believe this is ready to merge. all 813 tests pass locally. please note that minimum build compiler is now 8.10.1.

shayne-fletcher commented:
p.s. locally my testing was with `refactor-0.9.2.0`

shayne-fletcher commented:
> @shayne-fletcher Removing that test for now is good. Once you make cabal build work for this patch, I can look into it - otherwise it's difficult for me to investigate.

@zliu41 `cabal v2-build` on this branch works as you'd hope for now!

zliu41 commented on src/Hint/Extensions.hs:
```diff
@@ -515,19 +515,21 @@ addDerives nt _ xs = mempty
   where
-    idecl :: DataFamInstDecl GhcPs -> Derives
-    idecl (DataFamInstDecl (HsIB _ FamEqn {feqn_rhs=HsDataDefn {dd_ND=dn, dd_derivs=(L _ ds)}})) = g dn ds
+    idecl :: LocatedA (DataFamInstDecl GhcPs) -> Derives
```
You should remove the `LocatedA` here. Then the test should pass.

shayne-fletcher replied:
outstanding @zliu41 , many thanks. i have restored the (now passing) test!

ndmitchell commented:
Thanks! @shayne-fletcher  - can you add a CI for GHC 9.2 and remove it for GHC 8.8 (see the `.github/workflows`). 

It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

I'll take a peek through the code now

ndmitchell commented on src/GHC/Util/SrcLoc.hs:
```diff
@@ -25,6 +37,12 @@ newtype SrcSpanD = SrcSpanD SrcSpan
   deriving (Outputable, Eq)
 instance Default SrcSpanD where def = SrcSpanD noSrcSpan
 
+newtype FastStringD = FastStringD FastString
```
Does GHC not provide an Ord for this type? Could we get one added upstream for the future?

shayne-fletcher replied:
@ndmitchell  the `Ord` instance has been removed. see https://gitlab.haskell.org/ghc/ghc/-/commit/4b4fbc58d37d37457144014ef82bdd928de175df. however i overlooked the existence of `LexicalFastString` (see https://hackage.haskell.org/package/ghc-9.2.1/docs/GHC-Data-FastString.html#t:LexicalFastString) which we can use here. pushing a fix for that now.

ndmitchell reviewed (commented):
Nice work!

shayne-fletcher commented:
> Thanks! @shayne-fletcher - can you add a CI for GHC 9.2 and remove it for GHC 8.8 (see the `.github/workflows`).

pushed

> It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

no idea about that sorry.

> I'll take a peek through the code now

alanz commented:
> It seems the Windows and Mac CI are both bust installing apply-refact. Any idea why?

It depends on ghc-exactprint, the release on hackage does not compile with GHC 9.2.1 as released.

I will be making a new release tomorrow,  just need to clean up a few tests

ndmitchell commented:
Thanks @alanz - my plan is to hold this PR in limbo a few days, merge nothing else to the repo in the intervening time, and wait for ghc-exactprint, then pull the trigger on this.

alanz commented:
I just published [ghc-exactprint-1.3.0](https://hackage.haskell.org/package/ghc-exactprint-1.3.0)

arjunkathuria commented:
hi, if all's good can we merge this ? looking forward : )

unclechu commented on stack.yaml:
```diff
@@ -1,19 +1,23 @@
-#   ghc-lib-parser-ex:
-#     auto: false
-#     no-ghc-lib: false
+#  hlint:
+#    ghc-lib: true
+#  ghc-lib-parser-ex:
+#    auto: false
+#    no-ghc-lib: false
```
```suggestion
#   hlint:
#     ghc-lib: true
#   ghc-lib-parser-ex:
#     auto: false
#     no-ghc-lib: false
```

unclechu commented on src/Main.hs:
```diff
@@ -8,7 +8,7 @@ import System.Exit
-    errs <- hlint args
-    unless (null errs) $
-        exitWith $ ExitFailure 1
+     args <- getArgs
+     errs <- hlint args
+     unless (null errs) $
+         exitWith $ ExitFailure 1
```
5 spaces for indentation? But still the `exitWith` has 4.

```suggestion
    args <- getArgs
    errs <- hlint args
    unless (null errs) $
        exitWith $ ExitFailure 1
```

unclechu commented on src/Hint/Restrict.hs:
```diff
@@ -181,14 +180,15 @@ importListToIdents =
-                         IEType    n -> ("type " ++) <$> fromId (unLoc n)
+    fromName wrapped =
+      case unLoc wrapped of
+        IEName      n   -> fromId (unLoc n)
+        IEPattern _ n -> ("pattern " ++) <$> fromId (unLoc n)
+        IEType    _ n-> ("type " ++) <$> fromId (unLoc n)
```
Broken alignment:
```suggestion
        IEName      n -> fromId (unLoc n)
        IEPattern _ n -> ("pattern " ++) <$> fromId (unLoc n)
        IEType    _ n -> ("type " ++) <$> fromId (unLoc n)
```

unclechu commented on src/GHC/Util/Scope.hs:
```diff
@@ -38,11 +38,13 @@ scopeCreate xs = Scope $ [prelude | not $ any isPrelude res] ++ res
     -- The import declaraions contained by the module 'xs'.
     res :: [LImportDecl GhcPs]
-    res = [x | x <- hsmodImports xs , pkg x /= Just (StringLiteral NoSourceText (fsLit "hint"))]
+    res = [x | x <- hsmodImports xs
+              , pkg x /= Just (StringLiteral NoSourceText (fsLit "hint") Nothing)
+            ]
```
```suggestion
    res = [x | x <- hsmodImports xs, pkg x /= Just (StringLiteral NoSourceText (fsLit "hint") Nothing)]
```

ndmitchell commented:
It's got to the stage where we need to move, so landing, and will fix forward if we need to.

ndmitchell commented:
Merged, but all the CI is red on GHC 9.2 because apply-refact is not yet GHC 9.2 compatiblehttps://github.com/mpickering/apply-refact/issues/114

alanz commented:
I am working on the failing tests for apply-refact, and came across `Bracket7`, which has

```haskell
yes = if (f x) then y else z
```
and a corresponding apply-refact refactoring of
```haskell
[("tests/examples/Bracket7.hs:1:10-14: Suggestion: Redundant bracket\nFound:\n  if (f x) then y else z\nPerhaps:\n  if f x then y else z\n"
,[Replace
  { rtype = Expr
  , pos = SrcSpan {startLine = 1, startCol = 10, endLine = 1, endCol = 15}
  , subts = [("x",SrcSpan {startLine = 1, startCol = 11, endLine = 1, endCol = 14})]
  , orig = "x"}])]
```
I presume that only the positions are meaningful, as in "substitute 1:10-15 with 1:11-14", and the reference to `"x"` can be ignored?

ndmitchell commented:
@alanz  - @zliu41 is the expert here. I think (but am most definitely not sure) that the code is saying:

* Due to `pos` Operate on 1:10-15, i.e. `(f x)`
* Due to `subts` bind `$x` to 1:11-14, i.e. `f x`
* Replace the whole thing with `orig` definition of `$x` so replace `(f x)` with `f x`

But that's just a guess on what I think it's doing.

alanz commented:
That seems to make sense. So the "x" is a "hlint variable", rather than a quote of existing source code.

zliu41 commented:
@alanz Yeah "x" is a template variable. In this case it happens to be the entire template (`orig`), but in general that's not the case.

alanz commented:
https://github.com/mpickering/apply-refact/pull/116 is ready for review, should hopefully unblock this
